### PR TITLE
fix(log/stream): ensure the appender future always gets completed

### LINKED ISSUE #7992: OutOfMemory in follower
**Describe the bug**

In the release-1.2.0 benchmark, observed that the node which is the follower for all partitions went out of memory. 
Might be related to #7784 

image

https://console.cloud.google.com/errors/CMz-q-2yh8X-pQE?service=zeebe-broker&time=P7D&project=zeebe-io

**Environment:**
 - Zeebe Version: 1.2.0

### PR DESCRIPTION
**Pull request by romansmirnov**
## Description

* Handles any kind of thrown `Throwable`s in the `LogStream` actor, so that the appender future gets completed exceptionally.
* Handles the situation when opening the appender, the `LogStream` actor is supposed to be closed. In this situation, the appender future gets completed exceptionally as well.

## Related issues

closes #7992

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [x] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStreamImpl.java [hunk 1]
```diff
@@ -382,4 +403,11 @@ public void onUnrecoverableFailure(final HealthReport report) {
 
     T create(int partitionId, Dispatcher dispatcher);
   }
+
+  private static final class ShouldCloseAppenderException extends RuntimeException {
```
Comment (oleschoenburg) [inline logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStreamImpl.java:407]: :wrench: Can we rename this to something like `LogStorageAppenderClosed`? When this exception appears, the appender is already closed so I think "shouldClose" is confusing. Similarly, I think the exception message could be something along the lines of "LogStorageAppender was closed before opening succeeded", the important part being that a close wasn't just requested but has already happened.

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStreamImpl.java [hunk 2]
```diff
@@ -135,11 +136,25 @@ public void close() {
 
   @Override
   protected void handleFailure(final Throwable failure) {
```
Comment (oleschoenburg) [inline logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStreamImpl.java:138]: :wrench: Could we extract an internal method that we can then call from `openAppender`?  For me it's a bit confusing that we are calling an actor method manually, `LogStreamImpl` would be the only class where we do this.

**Review by lenaschoenburg — APPROVED** [review]
Thanks @romansmirnov :+1: 

I have two minor :wrench: comments but nothing major that would require a re-review.

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStreamImpl.java (hunk 1 above)
Comment (romansmirnov) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStreamImpl.java]: Done

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStreamImpl.java (hunk 2 above)
Comment (romansmirnov) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStreamImpl.java:138]: Done

**Comment by romansmirnov** [comment]
bors retry

**Pull request merged**